### ISSUE TITLE
2 users got the same session? 🤗 

### ISSUE THREAD

**Issue by G2G2G2G** [gofiber/fiber#1497]
I can't think of how else this happened:
one user today had their name switching in the logs, randomly it'd be them, then the other person. is it possible both got signed the same UUID? 

What is the correct way to set a session, is this wrong?
**Do I need to check every other session to make sure it isn't the same as someone else?**
Do I need to have more than 1 cookie set to check against? 2 sessions or something? (I know a lot of sites have something like that) 

```go
sess, err := ses.Get(c)
	if err != nil {
		log.Println("login.go: SES: ", err)
		return c.SendString("Session Error 1")
	}
	defer sess.Save()

	if err := sess.Regenerate(); err != nil {
		log.Println("login.go: sesRegen: ", err)
		return c.SendString("Session Error 2")
	}
//pull a lot of stuff from database

// this is the variable that kept switching in the database logs, all of it is server side but it shows that person switching this variable randomly
sess.Set("user", login)   
//set the rest of database stuff same way as above

```

Method used to input that user's name in the database logs (where it shows them switching between users randomly throughout the day): 

```go

func getSesUser(c *fiber.Ctx) string {
	sess, err := ses.Get(c)
	if err != nil {
		return ""
	}
	user := sess.Get("user")
	if user != nil {
		return user.(string)
	}
	return ""
}
```

**Comment by hi019** [comment]
It looks fine. Can you provide a minimal reproduction of the issue?

**Comment by G2G2G2G** [comment]
I don't think so. As I said I think it generated the same UUID for 2 users, the chances of that aren't even possible to reproduce?

I did a few things to hopefully help this:
- made my key 80 characters long instead of the uuid (with https://golang.org/pkg/crypto/rand/ )
- I change the key every time page is refreshed
- I set a second cookie with their logged in name, and match that to the session key in the backend every time they refresh, if their username changed (as if they got the same session) it logs both of the people out

Doing that I can no longer copy the session ID from one browser and then become them in a different browser (basic session hijacking) 
I think it won't happen anymore, I realize the chances of this happening again are probably close to zero with only 10 employees using this system.. but it did happen once which is crazy enough right?

I don't know enough about this at all to know what to do, so I did several things as seen above..

**Comment by hi019** [comment]
Sorry I missed your questions. You shouldn't need to check every session, or have more than one. Assuming the problem is with our [UUID generator](https://github.com/gofiber/fiber/blob/861e5d21fbdfcf1759a274a3becfa5d9cf9a173c/utils/common.go#L70), using crypto/rand will suffice. But we still need to make sure your code is OK. Are you using the session instance after you call session.Save?

**Comment by ReneWerner87** [comment]
i tested the UUID && UUID_v4 generation with 10000000 ids in parallel goroutines and with random delay, none of the generated ids was identical


![image](https://user-images.githubusercontent.com/7063188/130410668-77d6f5e7-21c9-417d-9646-7873b7a58d76.png)

have tested both functions in other constellations, unparallel for example

don't think it's possible that the uuid is identical, with several runs with over 10 million ids per run, i couldn't find any error with this generation

@G2G2G2G so the error is that 2 users got identical sessions ... how often did it happen ? was this a case of session capturing ? can you reproduce it ?

we can only help if we have a small example with reproducible code, from our tests, it showed that it is not possible, unless we are talking about the known hacker attacks, which of course should be prevented by following the guides from the owasp

https://owasp.org/www-community/attacks/Session_hijacking_attack
https://owasp.org/www-community/attacks/Session_fixation
https://owasp.org/www-community/attacks/Session_Prediction

**Comment by G2G2G2G** [comment]
@hi019 I read from the session if that's what you mean by using the session instance?
I read from it many times, here is my login function:
```Go

func login(c *fiber.Ctx) error {
	p := new(Person)
	ip := getIPstring(c)
	if err := c.BodyParser(p); err != nil {
		log.Println("LOGIN PARSER: ", err)
		return c.SendString("Error parsing input.")
	}

	//check lengths
	if len(p.User) < 2 || len(p.Pin) < 2 || len(p.Pass) < 2 {
		return c.SendString("Input some info!")
	}

	sess, err := ses.Get(c)
	if err != nil {
		log.Println("login.go: SES: ", err)
		return c.SendString("Session Error 1")
	}
	defer sess.Save()

	if err := sess.Regenerate(); err != nil {
		log.Println("login.go: sesRegen: ", err)
		return c.SendString("Session Error 2")
	}

	//allow users that are on the office LAN to login without having to specify password. Only user/pin
	if ip != "10.0.35.1" && p.Pass != "removed 200 character key" {
		if !checkPass(p.User, p.Pass) {
			go inputLogs(p.User, ip, "LOGIN", "0", "0", `{"type":"LOGIN","data":"PASS FAIL"}`)
			return c.SendString("WRONG")
		}
	}

	var login string
	var users, hours, estimates, backup, graphs, gallery, logs, estimatesEmail, homePanelProg, chpass bool
	err = db.QueryRow("SELECT login,users,hours,estimates,backup,graphs,gallery,logs,estimates_email,home_panel_prog,chpass FROM users WHERE login = $1 AND pin = $2 AND archived = false", p.User, p.Pin).Scan(&login, &users, &hours, &estimates, &backup, &graphs, &gallery, &logs, &estimatesEmail, &homePanelProg, &chpass)
	if err != nil {
		go inputLogs(p.User, ip, "LOGIN", "0", "0", `{"type":"LOGIN","data":"PIN FAIL"}`)
		return c.SendString("WRONG")
	}

	cookie := new(fiber.Cookie)
	cookie.Name = "USER"
	cookie.Value = login
	cookie.HTTPOnly = true
	cookie.Secure = true
	cookie.Expires = time.Now().Add(3 * 24 * time.Hour)
	c.Cookie(cookie)

	if chpass {
		//chpass is true if admin randomly generated a new password for someone who forgot
		//check if chpass is over the 5 days and don't let them login so users are required to change passwords after they forget

		//return c.SendString("Temporary pass expired")
	}

	go inputLogs(login, ip, "LOGIN", "1", "0", `{"type":"LOGIN","data":"SUCCESS"}`)
	sess.Set("user", login)
	sess.Set("hours", hours)
	sess.Set("users", users)
	sess.Set("estimates", estimates)
	sess.Set("backup", backup)
	sess.Set("graphs", graphs)
	sess.Set("gallery", gallery)
	sess.Set("logs", logs)
	sess.Set("estimates_email", estimatesEmail)
	sess.Set("home_panel_prog", homePanelProg)
	sess.Set("chpass", chpass)
	return c.SendString("OK")
}
```
All of the "sess.Set" stuff I use on various pages

The COOKIE above is new, I have this upon loading the page:
```Go
if user != c.Cookies("USER") {
			log.Println("SESSION ERROR:", user, "!=", c.Cookies("USER"), getSesID(c))
			return logoutUser(c)
		}
```
Which is a method I've read about, to kick users out even if the session gets hijacked, the hijacker would need to copy both cookies..**. (not an issue we have, our login is private (mostly in LAN) and access logs show nobody has ever visited in the past year)** 


Here is how I get their user on all other pages:
```Go
func getSesUser(c *fiber.Ctx) string {
	sess, err := ses.Get(c)
	if err != nil {
		return ""
	}
	user := sess.Get("user")
	if user != nil {
		return user.(string)
	}
	return ""
}
```
Here is how I check whether or not they have auth for certain things:
```Go

func checkAuth(c *fiber.Ctx, page string) bool {
	sess, err := ses.Get(c)
	if err != nil {
		return false
	}
	if sess.Get("user") != nil {
		if sess.Get(page) == true {
			return true
		}
	}
	return false
}
```

I never called save ever again (except now that I regenerate a new key every refresh, but previously I did not)


@ReneWerner87 
"How often did it happen"
once in the past year that we know of. No I can't reproduce it as I said lol. I've tried what you did, but as far as I can see it's so rare you can't reproduce it? idk enough about it.

I'm using the INMEMORY sessions. From what I've read people with databases are supposed to set the keys to UNIQUE so nobody gets the same thing.. 


From my logs and the user's cellphone capturing their office computer showing me they had authorization of a different user, 
- **they refreshed the page and their user was switching between them and another user** 
- it seemed a 50/50 chance which brought me to my question of whether or not they had the same ID and that's what did it? 
- ## I don't even know if they DID have the same ID, all I can think of is that's what happened though
as I posted in my OP I just asked a bunch of questions, I don't know how it happened, I just know it did happen and all I can think of is they had the same identification process through my `getSesUser` function which somehow switched them between users and that user's full authentication.

I am 100% sure it was this other user, as our user names are shown on the top of the web page. and it switched back and forth from them. 

When a user refreshes the page, the one page they get is the menu page which is uniquely generated to their user, here is the part of that view:
```html
<div class="menu" id="mainmenu"><a id="menuh" href="/aDMiN_Panel/#h">Home</a>
{{unescape .MenuItems}}
<a id="menul" href="/aDMiN_Panel/#l">Logs</a><a href="#logout">Logout of {{.User}}</a>
```
Their {{.User}} is set with the `getSesUser` and the MenuItems are set by checking each of the ` checkAuth` variables to see their "access"
All of this is also checked upon accessing that content as well!!! but their menus are set so if they don't have access they don't have to see the extra menus for different sections either. 

Once again somehow upon refreshing (I don't know how because the user sent a bad phone example of it happening then shut down their computer) their ENTIRE session changed, including the output of `getSesUser` which reflected another user, they had all of their menu items etc. Effectively being that other user entirely. The front end view of their users etc is never used for auth, only the `getSesUser` and `checkAuth` are. Which are set ONE TIME in all of my code (in that login)

>we can only help if we have a small example with reproducible code, from our tests, it showed that it is not possible, unless we are talking about the known hacker attacks, which of course should be prevented by following the guides from the owasp

Yea I know that, I've tested it thoroughly and can't reproduce it either. All I can see is somehow this function:
https://github.com/gofiber/fiber/blob/v2.17.0/middleware/session/store.go#L39
Got someone elses session, with only 1 browser cookie set (the session one) it switched between users, which is why I think it was something to do with the session ID but I'm not 100% sure
### Can anyone think of any other way it'd have gotten someone else's identify via the Get function? From the looks of it, it gets `getSessionID` which is expected.

If there's no ideas or I am wrong somewhere and something needs to be fixed or anything I guess this can be closed. I told them to never logout like that again so I can test stuff myself. I don't think it'll happen ever again though since now I have a static cookie of their user in the browser and if they did somehow switch users it'll kick them out of the login session.

**Comment by G2G2G2G** [comment]
Other than their phone stuff which shows the user change, here is what are in my logs with some info, so you can see what I see which.. is weird I don't know how it can happen as I said many times, it is really confusing:

Trying to be as confidential here as possible sorry, removed the IP except last octet etc
no logout and then login was involved, between when they switched users, it was completely random as my logs show (my logs only show when someone edits something on one of our client pages)
(note the page column will always say LOGIN (cuz it is that page) but the data shows if it was a login or logout)
**This stuff is in descending order so read from bottom to top**


```SQL
select * from logs where epoch > 1629290331 and epoch < 1629407940 and loginname != 'OMITTED' and loginname != 'OMITTED' order by epoch desc;
   epoch    | loginname |   ipaddress    |   page    |  id  |                              data                              | estimateID 
------------+-----------+----------------+-----------+------+----------------------------------------------------------------+---------
 1629407662 | jREMOVED      | IPREMOVED.223 | LOGIN     |    1 | {"type":"LOGOUT","data":"SUCCESS"}                              |       0
 1629407354 | jREMOVED      | IPREMOVED.223 | LOGIN     |    1 | {"type":"LOGIN","data":"SUCCESS"}                               |       0

 //user logged out but upon that logout, it showed them as the other user, once again
 //I theorize because it was a 50/50 chance to be that other user, and the last few chances just didn't roll the dice well
 1629407303 | gREMOVED      | IPREMOVED.223 | LOGIN     |    1 | {"type":"LOGOUT","data":"SUCCESS"}                              |       0
 1629397321 | jREMOVED      | IPREMOVED.223 | ESTIMATES | 3206 | {"col":"artstatus","data":"Ready For Review"}                   |    3206
 1629397317 | jREMOVED      | IPREMOVED.223 | ESTIMATES | 3206 | {"col":"notes","data":"lots of stuff..."}                       |    3206
 1629397166 | jREMOVED      | IPREMOVED.223 | ESTIMATES | 3104 | {"col":"notes","data":"lots of stuff..."}                       |    3104

 //user switched back to themselves somehow
 1629397133 | jREMOVED      | IPREMOVED.223 | ESTIMATES | 3104 | {"col":"artstatus","data":"Ready For Review"}                   |    3104

 //user switched to other user somehow? Same IP address, this is their HOME ADDRESS, nobody else was there.
 1629397078 | gREMOVED      | IPREMOVED.223 | ESTIMATES | 3337 | {"col":"artstatus","data":"Ready For Review"}                   |    3337
 
 //user logged in
 1629394930 | jREMOVED      | IPREMOVED.223 | LOGIN     |    1 | {"type":"LOGIN","data":"SUCCESS"}                               |       0
```

These log names are entered via:
```Go
func getSesUser(c *fiber.Ctx) string {
	sess, err := ses.Get(c)
	if err != nil {
		return ""
	}
	user := sess.Get("user")
	if user != nil {
		return user.(string)
	}
	return ""
}
```
So as we see the names changed from "j" and "g" 

```Go
func inputLogs(name, ip, page, id, estimateID, data string) {
	ut := time.Now().Unix()
	_, err := db.Exec("INSERT INTO logs (loginname, ipaddress, page, id, estimateID, data, epoch) VALUES($1,$2,$3,$4,$5,$6,$7)", name, ip, page, id, estimateID, data, ut)
	if err != nil {
		log.Println("inputLogs: ", err)
		return
	}
}
```


# can the web socket somehow clash with it and change a user?

```Go

func upgradeSocket(c *fiber.Ctx) error {
       //checks if user is allowed to be logged in, all people must have logs access
	if !checkAuth(c, "logs") {
		return c.Status(403).SendString("Nope")
	}
	if websocket.IsWebSocketUpgrade(c) {
		if c.Get("host") == "OMITTED" {
			c.Locals("user", getSesUser(c))
			c.Locals("ip", getIPstring(c))
			c.Locals("users", checkAuth(c, "users"))
			c.Locals("hours", checkAuth(c, "hours"))
			c.Locals("estimates", checkAuth(c, "estimates"))
			c.Locals("estimates_email", checkAuth(c, "estimates_email"))
			c.Locals("home_panel_prog", checkAuth(c, "home_panel_prog"))
			return c.Next()
		}
	}
	return fiber.ErrUpgradeRequired
}
```
Socket does not save session info, it just reads from it.. I don't even know how to go back to the regular `c *fiber.Ctx` thing, it just sends those variables to the person's websocket. 

I read it like this:
```Go
fmt.Sprint(c.conn.Locals("user"))
or
c.conn.Locals("estimates_email").(bool)
```

**Comment by ReneWerner87** [comment]
Okay thanks for this information, if it were a bug it would occur more than once a year and also with others

Since you don't know for sure if it's the session logic and here is a complex extension for the login, the whole problem could be there as well

I guess there is only someone logged out within the 3 days where the cookie runs and logged in with another user login or changed the user directly .... There I can imagine some scenarios http and https (secured, http only)

Since this bug occurred only once, we can't help because it's not reproducible, additionally we can't exclude errors in the logic which is in the microservice or in the process of the user handling, I would like to close the bug, because I think we can't help any further

**Comment by G2G2G2G** [comment]
>I guess there is only someone logged out within the 3 days where the cookie runs and logged in with another user login or changed the user directly

Yea I ran into this many months ago which was fixed by you or hi019's (I forgot) help via:
```Go
if err := sess.Regenerate(); err != nil {
		log.Println("login.go: sesRegen: ", err)
		return c.SendString("Session Error 2")
	}
```
(you can see it in my login function) without regenning it, as you said, the cookie stays the same and did cause the same issue. But that was fixed (maybe even last year? I forgot, but I know I asked for help there too) 


I'll close it, it's good to know that what I have seems correct at least, If it ever happens again I now log the sessions generated in log.Println so I'll maybe have a better idea if someone notices it happening (the office employees don't usually pay attention to that stuff lol)
Thanks guys

**Comment by G2G2G2G** [comment]
@ReneWerner87 I never figured this out, however today I watched a video where http1/1 has some bug that has people switching sessions with other people on big sites as well.

https://youtu.be/n3Bw8CASnHE

Not sure it is related and it is beyond me.
My go program was behind nginx which of course notoriously is a reverse proxy that only has http1/1 until relatively recently 
https://serverfault.com/questions/765258/use-http-2-0-between-nginx-reverse-proxy-and-backend-webserver

I wrote some other code to kick people out if their ID doesn't match the local storage ID I generate and it has only triggered a few times last several years, and hasn't for at least 2 years now. So maybe nginx fixed it or maybe debian did or maybe somewhere in Go it did, who knows.
